### PR TITLE
Show progress while fetching dataset values

### LINKED ISSUE #892: Progress bar
Like on Braggy, we could show a progress bar on the `ValueLoader`, for dataset values requests.

### PR DESCRIPTION
**Pull request by axelboc**
Fix #892 

New version with listeners in the API.

Still the problem that the back-end doesn't send the total size of the data for some reason (probably missing `Content-Length` HTTP header).

### CODE REVIEW

**Review by loichuder — DISMISSED** [review]
:partying_face: 

I will have a look at the back-end

File: packages/app/src/vis-packs/ValueLoader.tsx
Comment (loichuder) [inline packages/app/src/vis-packs/ValueLoader.tsx:44]: I tried by adding the `Content-Length` headers in h5grove and it works !

I would simply suggest to round `progress` as now, it displays a number with many decimals.

Also, the `ValueLoader` displays `0%` just before disappearing which is not great. Perhaps due to the `removeProgressListener` ?

Comment (axelboc) [reply packages/app/src/vis-packs/ValueLoader.tsx:44]: Great! Okay, I've rounded the number and I've made sure the `progress` is reset when a new request starts rather than when a request fulfills.

**Comment by axelboc** [comment]
I'll wait until you've merged/deployed the back-end on bosquet. I'd like to double check what happens when multiple value requests occur in parallel (i.e. with NeXus).

**Comment by axelboc** [comment]
Alright, I've pushed something I'm starting to feel okay with. I now display progress bars for ongoing requests:

image

If multiple requests are happening in parallel (e.g. NeXus), the `ValueLoader` shows as many progress bars, stacked on top of one another. The bars disappear when the corresponding requests fulfill.

I'm still not 100% happy with the code, but at least having multiple progress bars makes sense conceptually and means we can keep everything decoupled (i.e. the `ValueLoader` doesn't need to know which requests are in progress; we don't need to initialize a `progress` state in every container; we don't need to pass a callback from every call to `useDatasetValue(s)`; etc.)

One problem I see is that since we don't cancel requests when slicing, the progress bars add up:

image

I think it's worth cancelling requests like we do when switching vis/entity -- if we don't want to do that, then we'll be back in a pickle :smile: 

Anyways... to be continued.

**Comment by axelboc** [comment]
... or I can just show the five "oldest" progress bars. Once the fetch of a previous slice succeeds, the progress bar disappears and leaves room for a more recent progress bar. Sounds reasonable to you @loichuder?

**Comment by loichuder** [comment]
That sounds sensible. I will wait to test it out to give my final opinion :smile:

**Comment by axelboc** [comment]
Okay, I've limited it to 3 simultaneous progress bars. Enough to show that multiple things are loaded, yet not too many so as for the UI to not feel as cluttered as in the previous screenshot. It works for NeXus groups with multiple children datasets to fetch, as well as when slicing. We may decide to cancel requests while fetching, as previously discussed (and as suggested in #920), but with the new 3-bars limit, this no longer impacts this PR.

File: packages/app/src/providers/api.ts
Comment (axelboc) [inline packages/app/src/providers/api.ts:76]: Note a minor change here: `queryParams` defaulting to `storeParams` was prone to bugs in my opinion. We no longer relied on it anyway.

File: packages/app/src/providers/api.ts
Comment (loichuder) [inline packages/app/src/providers/api.ts:107]: Just a thought, if `progressListeners` was a `Map` instead of a `Set`, you could notify only the relevant change, right ?

You would need to pass the `storeParams` as arg of most callbacks though.

File: packages/app/src/vis-packs/ValueLoader.module.css
Comment (loichuder) [inline packages/app/src/vis-packs/ValueLoader.module.css:14]: I would change the CSS (add a `min-height` ?) so that the `Cancel` button is not moved around when progress bars disappear. 

It might be frustrating for the user to try to click on the cancel button if the layout changes dynamically...

**Review by loichuder — APPROVED** [review]
Nice :+1:

File: packages/app/src/providers/api.ts
Comment (axelboc) [reply packages/app/src/providers/api.ts:107]: You mean so listeners only get notified of the request that has "progressed"? Not sure there's a need for this. The value loader loops through all the progress values on every render anyway. We don't have a component that would benefit from re-rendering only when a specific request's progress changes.

File: packages/app/src/vis-packs/ValueLoader.module.css
Comment (axelboc) [reply packages/app/src/vis-packs/ValueLoader.module.css:16]: Good call. I've added a `min-height` as you suggested, as well as `justify-content: center` so the whitespace is shared equally above/below the bar(s) when there's only one or two.

**Pull request merged**
